File: python/sameTree.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None) -> None:
        self.val = val
        self.left = left
        self.right = right


def isSameTree(p, q) -> bool:
    return recurse(p, q)


def recurse(p, q) -> bool:
    if p is None and q is None:
        return True
    elif p is None or q is None:
        return False

    if p.val != q.val:
        return False

    return recurse(p.left, q.left) and recurse(p.right, q.right)

File: go/SameTree/sameTree.go
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

// Time -> 0(N) Space -> 0(log(N))
func isSameTree(p *TreeNode, q *TreeNode) bool {
	return recurse(p, q)
}

func recurse(p, q *TreeNode) bool {
	if p == nil && q == nil {
		return true
	} else if p == nil || q == nil {
		return false
	}

	if p.Val != q.Val {
		return false
	}

	return recurse(p.Left, q.Left) && recurse(p.Right, q.Right)
}
